packages/react-jss/src/createUseStyles.js

```diff
@@ -0,0 +1,148 @@
+import {manageSheet, unmanageSheet} from './utils/managers'
+import {getSheetClasses} from './utils/getSheetClasses'
+
+const noTheme = {}
+
+const createUseStyles = <Theme: {}>(styles: Styles<Theme>, options?: HookOptions<Theme> = {}) => {
+  const {index = getIndex(), theming, name = 'Hook', ...sheetOptions} = options
+  const isFirstMount = React.useRef(true)
+  const ThemeContext = (theming && theming.context) || DefaultThemeContext
+  const useTheme =
+    typeof styles === 'function'
+      ? // $FlowFixMe
+        (): Theme => React.useContext(ThemeContext)
+      : // $FlowFixMe
+        (): Theme => noTheme
+
+  const useLayoutEffect = isInBrowser ? React.useLayoutEffect : React.useEffect
+
+  return (data: any) => {
+    const context = React.useContext(JssContext)
+    const theme = useTheme()
+
+    if (context.disableStylesGeneration) {
+      return {}
+    }
+
+    const [staticSheet, setStaticSheet] = React.useState(() => {
+      const sheet = createStaticSheet({
+        context,
+        styles,
+        name,
+        theme,
+        index,
+        sheetOptions
+      })
+
+      if (context.registry && sheet) {
+        context.registry.add(sheet)
+      }
+
+      return sheet
+    })
+
+    const [dynamicRules, setDynamicRules] = React.useState(() => {
+      if (staticSheet) {
+        return addDynamicRules(staticSheet, data)
+      }
+
+      return undefined
+    })
+
+    const [classes, setClasses] = React.useState(() => {
+      if (staticSheet) {
+        return getSheetClasses(staticSheet, dynamicRules)
+      }
+
+      return {}
+    })
+
+    useLayoutEffect(
+      () => {
+        if (staticSheet) {
+          manageSheet({
+            index,
+            context,
+            sheet: staticSheet,
+            theme
+          })
+        }
+
+        return () => {
+          if (staticSheet) {
+            unmanageSheet({
+              index,
+              context,
+              sheet: staticSheet,
+              theme
+            })
+
+            if (dynamicRules) {
+              removeDynamicRules(staticSheet, dynamicRules)
+            }
+          }
+        }
+      },
+      [staticSheet]
+    )
+
+    useLayoutEffect(
+      () => {
+        if (dynamicRules && staticSheet) {
+          updateDynamicRules(data, staticSheet, dynamicRules)
+        }
+      },
+      [data]
+    )
+
+    useLayoutEffect(
+      () => {
+        if (!isFirstMount.current) {
+          const newStaticSheet = createStaticSheet({
+            context,
+            styles,
+            name,
+            theme,
+            index,
+            sheetOptions
+          })
+          const newDynamicRules = staticSheet ? addDynamicRules(staticSheet, data) : undefined
+          const newClasses = staticSheet ? getSheetClasses(staticSheet, dynamicRules) : undefined
+
+          setStaticSheet(newStaticSheet)
+          setDynamicRules(newDynamicRules)
+          setClasses(newClasses)
```

Comment:
Are these batched or can each call trigger a separate render? If so can this be an issue? It looks like those should live in a single state i.e. use `useReducer`.

Reply:
The ` setState` calls are batched from react AFAIK. But we could also change whole implementation to a `useReducer` call.

Reply:
setState is batched since quite some time in react, you can call any amount of setState, the state is getting merged and scheduled

Reply:
It's batched in event handlers but not in general (which is why `unstable_batchedUpdates` exists). Don't know about effects though. Does anybody have a resource comparing setState behavior in render, layout effects, passive effects and event handlers?

https://github.com/facebook/react/issues/10231#issuecomment-316644950 makes me question that they are indeed batched. Link is old though and may very well not apply anymore.

I would expect react to batch anything that it invokes itself (event handlers, renders, effects etc). So the code would be fine.